### ISSUE TITLE
[data grid] Bug with selection, row grouping and rowSelectionPropagation

### ISSUE THREAD

**Issue by M-sleeper** [mui/mui-x#15753]
### Steps to reproduce

Steps:
1. Open this link to live example: https://codesandbox.io/p/sandbox/mystifying-panini-m4ghxj
2. wait 2 seconds after the table is loaded
3. Check checkbox of "James Cameron" in the group of "20th Century Fox"

### Current behavior

All the children of group "James Cameron" are selected, the group's checkbox is in the intermediate state.

### Expected behavior

Having
```
rowSelectionPropagation={{ parents: true, descendants: true }}
```
I would expect the checkbox of group "James Cameron" is selected, when all its children are (programmatically) selected. The docs says:
> Selecting all the filtered descendants of a parent selects the parent automatically

https://mui.com/x/react-data-grid/row-grouping/#automatic-parents-and-children-selection 

Deselecting and selecting one of the children makes the parent selected as expected.

### Context

Related to https://github.com/mui/mui-x/issues/15382

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  System:
    OS: Linux 6.8 Ubuntu 24.04.1 LTS 24.04.1 LTS (Noble Numbat)
  Binaries:
    Node: 22.12.0 - /usr/local/bin/node
    npm: 10.9.0 - /usr/local/bin/npm
    pnpm: Not Found
  Browsers:
    Chrome: 131.0.6778.108
  npmPackages:
    @emotion/react: ^11.13.3 => 11.13.3 
    @emotion/styled: ^11.13.0 => 11.13.0 
    @mui/core-downloads-tracker:  6.1.6 
    @mui/icons-material: ^6.1.6 => 6.1.6 
    @mui/material: ^6.1.6 => 6.1.6 
    @mui/private-theming:  6.1.6 
    @mui/styled-engine:  6.1.6 
    @mui/system:  6.1.6 
    @mui/types:  7.2.19 
    @mui/utils:  6.1.6 
    @mui/x-charts: ^7.22.1 => 7.22.1 
    @mui/x-charts-vendor:  7.20.0 
    @mui/x-data-grid:  7.23.0 
    @mui/x-data-grid-generator:  7.23.0 
    @mui/x-data-grid-premium: ^7.23.0 => 7.23.0 
    @mui/x-data-grid-pro:  7.23.0 
    @mui/x-date-pickers:  7.22.1 
    @mui/x-date-pickers-pro: ^7.22.1 => 7.22.1 
    @mui/x-internals:  7.21.0 
    @mui/x-license:  7.23.0 
    @mui/x-tree-view: ^7.22.1 => 7.22.1 
    @types/react:  18.3.12 
    react: 18.2.0 => 18.2.0 
    react-dom: 18.2.0 => 18.2.0 

```
</details>

**Search keywords**: rowSelectionPropagation datagrid row grouping
**Order ID**: 101072

**Comment by michelengelen** [comment]
Hey @M-sleeper ... It seems @MBilalShafi is taking this on already! 👍🏼

**Comment by M-sleeper** [comment]
@michelengelen thanks.

I found a probably also related bug:
1. Open live example here: https://codesandbox.io/p/sandbox/recursing-dawn-m8jddt?file=%2Fsrc%2FDemo.tsx%3A24%2C1
2. wait two seconds
3. there's an exception

Here I'm loading the rows later than the selection, using the `keepNonExistentRowsSelected` prop. I would expect that this should work properly: show the respective rows and eventually their parent as selected after the rows are loaded.

The use case is that we have a saved selection which is sooner available than the rows in the table which are loaded from the server when the component initially renders.

**Comment by M-sleeper** [comment]
@michelengelen Do you have any updates on this?

**Comment by michelengelen** [comment]
> @michelengelen Do you have any updates on this?

Not yet... @MBilalShafi is working on this afaik.

**Comment by MBilalShafi** [comment]
Correct, I own this bug fix, couldn't start actual work up until now since I've been busy with higher priority things, plus there's a significant [improvement](https://github.com/mui/mui-x/pull/15651) in the row selection feature under progress, I'd prefer to wait for it being merged before proceeding with the bug fix to avoid duplicate work.

**Comment by philicious** [comment]
I am encountering the same issue with `treeData` in combination with `rowSelectionModel`: 

I check all checkboxes by default like 
```jsx
  useEffect(() => {
    setRowSelectionModel(props.MyData.map((row) => row.id))
  }
  , [props.MyData])
```
but the parents show an intermediate state despite the grid having
```jsx
      rowSelectionPropagation={{
        parents: true,
        descendants: true
      }}
```

#15382 didnt fix it for treeData as it seems, only for rowGrouping
